Userland/Libraries/LibAudio/Metadata.cpp

```diff
@@ -0,0 +1,102 @@
+
+Optional<StringView> Person::name_for_role() const
+{
+    switch (this->role) {
+    case Role::Artist:
+        [[fallthrough]];
+    case Role::Performer:
+        return {};
+    case Role::Lyricist:
+        return "Lyricist"sv;
+    case Role::Conductor:
+        return "Conductor"sv;
+    case Role::Publisher:
+        return "Publisher"sv;
+    case Role::Engineer:
+        return "Engineer"sv;
+    case Role::Composer:
+        return "Composer"sv;
+    }
+    VERIFY_NOT_REACHED();
+}
+
+void Metadata::replace_encoder_with_serenity()
+{
+    auto version_or_error = Core::Version::read_long_version_string();
+    // Unset the encoder field in this case; we definitely want to replace the existing encoder field.
+    if (version_or_error.is_error())
+        this->encoder = {};
+    auto encoder_string = String::formatted("SerenityOS LibAudio {}", version_or_error.release_value());
+    if (encoder_string.is_error())
+        this->encoder = {};
+    this->encoder = encoder_string.release_value();
+}
+
+Optional<String> Metadata::first_artist() const
+{
+    return this->people.is_empty() ? Optional<String> {} : this->people.first().name;
+}
+
+ErrorOr<String> Metadata::all_artists(Optional<String> unknown_text_or_default, Optional<String> concatenate_with_or_default) const
+{
+    auto unknown_text = unknown_text_or_default.value_or(String::from_utf8_short_string("Unknown"sv));
+    auto concatenate_with = concatenate_with_or_default.value_or(String::from_utf8_short_string(", "sv));
+
+    // FIXME: This entire function could be similar to TRY(TRY(people.filter(...).try_map(...)).join(concatenate_with)).value_or(unknown_text) if these functional iterator transformers existed :^)
+    auto artists = this->people;
+    artists.remove_all_matching([](auto const& person) { return !person.is_artist(); });
+    if (artists.is_empty())
+        return unknown_text;
+
+    Vector<String> artist_texts;
+    TRY(artist_texts.try_ensure_capacity(artists.size()));
+    for (auto const& artist : artists) {
+        if (auto role_name = artist.name_for_role(); role_name.has_value())
+            artist_texts.unchecked_append(TRY(String::formatted("{} ({})", artist.name, role_name.release_value())));
+        else
+            artist_texts.unchecked_append(artist.name);
+    }
+    return String::join(concatenate_with, artist_texts);
+}
+
+Optional<Time> Metadata::normalized_time() const
+{
+    return time.visit(
+        [](Time const& time) -> Optional<Time> { return time; },
+        [](Year const& year) -> Optional<Time> { return Time::from_timestamp(static_cast<int>(year), 1, 1, 0, 0, 0, 0); },
+        [](Empty const&) -> Optional<Time> { return {}; });
+}
+
+ErrorOr<void> Metadata::add_miscellaneous(String const& field, String value)
+{
+    // FIXME: Since try_ensure does not return a reference to the contained value, we have to retrieve it separately.
+    (void)TRY(this->miscellaneous.try_ensure(field, []() { return Vector<String> {}; }));
```

Comment:
The initializer can't fail, so no need to use `try_ensure`. And the classic `ensure` returns a reference.

Reply:
The try_ensure has nothing to do with the initializer, `try_ensure` may allocate if it needs to grow the hash table.

Reply:
Oops, missed that part. But `try_ensure` ***has*** to do with the initializer fallibility tho